### ISSUE TITLE
Systeminformation.graphics This function will cause high CPU utilization

### ISSUE THREAD

**Issue by BlockCnFuture** [sebhildebrandt/systeminformation#932]
**Describe the bug**
When calling the Systeminformation.graphics function, especially when calling it within an interval, you will see high CPU utilization.

**To Reproduce**
1. used function 'Systeminformation.graphics' in Interval, call it every two second

**Current Output**
This function will cause high CPU utilization

**Expected behavior**
Should not cause high CPU utilization

**Environment (please complete the following information):**
 - systeminformation package version:  5.22.4
 - OS: Windows 11
 - Hardware HP15L

To get all needed environment information, please run the following command:

```
  ┌─────────────────────────────────────────────────────────────────────────────────────────┐
│  SYSTEMINFORMATION                                                      Version: 5.22.4 │
└─────────────────────────────────────────────────────────────────────────────────────────┘

Operating System:
──────────────────────────────────────────────────────────────────────────────────────────
Platform         : Windows
Distro           : Microsoft Windows 11 专业版
Release          : 10.0.22631
Codename         :
Kernel           : 10.0.22631
Arch             : x64
Hostname         : tuzhis
Codepage         : 936
Build            : 22631
Hypervisor       :
RemoteSession    : true

System:
──────────────────────────────────────────────────────────────────────────────────────────
Manufacturer     : HP
Model            : Victus by HP 15L Gaming Desktop TG02-2xxx
Version          :
Virtual          :

CPU:
──────────────────────────────────────────────────────────────────────────────────────────
Manufacturer     : Intel
Brand            : Core™ i7-14700F
Family           : 6
Model            : 183
Stepping         : 1
Speed            : 2.1
Cores            : 28
PhysicalCores    : 20
PerformanceCores : 28
EfficiencyCores  :
Processors       : 1
Socket           : LGA1700
```

**Additional context**
This problem is probably caused by using promise.all to execute multiple PowerShell commands concurrently under Windows. You can change it to execute one by one. The execution will not be much slower, but the CPU utilization problem should be optimized a lot.

**Comment by sebhildebrandt** [comment]
@BlockCnFuture Yes, on Windows the overhead calling Opening a PowerShell process is really high. I do not suggest calling this every 2 seconds. I already mentioned the whole situation in this issue: https://github.com/sebhildebrandt/systeminformation/issues/616

For version 6.0 I am working on something where we spin up a pool of PowerShell processes and they can then be used by each function ...

**Comment by TheRedfoox** [comment]
@sebhildebrandt Do you also not recommend calling every second graphics if `si.powerShellStart();` is used? What is currently the right way to monitor GPU metrics every second?

**Comment by sebhildebrandt** [comment]
@TheRedfoox You are right, even when using `si.powerShellStart()` this is to often. This function needs a few seconds to get everything for the underlying window commands. Unfortunately this is all much slower in windows that in macOS or Linux.

**Comment by TheRedfoox** [comment]
@sebhildebrandt I did not find a way to call only some GPU metrics (only GPU usage) that allowed to lighten the call. Do you confirm to me that it is not possible at the moment or did I miss something in the document?

**Comment by sebhildebrandt** [comment]
@TheRedfoox Yes I can confirm that (for now). The only way would be probably to write C++ or C# code ...

**Comment by mboudreau** [comment]
Yeah, I've just hit this myself which is frustrating as we really wanted to show GPU usage/temp to our users.

**Comment by mboudreau** [comment]
@sebhildebrandt After looking into this, I don't think you need powershell at all.  You can just use node's `child_process` to run `exec` on any command with pretty good performance.  For instance, running something like `node -e "const now = Date.now();require('child_process').exec('echo hello', () => console.log(Date.now()-now))"` will be returned in 21ms for myself.  You can actually run `nvidia-smi` this way and get it to output XML which can then be parsed into JSON.

[I found a package that already does this](https://github.com/quik-fe/node-nvidia-smi), which is what I'm using right now.  It's fully async and with my testing returns in about 250ms, which if you run the command directly, is about the same amount of time.

Unless absolutely necessary, I would remove all calls to using powershell for windows environments and simply use `exec`

**Comment by mboudreau** [comment]
Actually, scratch that, I had assumed it was because you used powershell to call nvidia-smi, but that doesn't seem to be the case.  You are using `exec`, but seem to be using `execSync` instead of the async version which would block the thread entirely....

**Comment by TheRedfoox** [comment]
@mboudreau Do you have a solution in the meantime ? I call graphics["*"] but as I refresh my data every second it generates a very slight freeze at the time of the call.

I really don't know what solution to tinker with, or even a clue

**Comment by mboudreau** [comment]
@TheRedfoox I ended up calling `nvidia-smi` directly from node, which should be on the PATH for any user with nvidia drivers installed:

```
export const getGPUs = async (): Promise<GPUObject> =>
	new Promise((resolve, reject) => {
		const query = [
			'index',
			'name',
			'pci.bus_id',
			'pci.sub_device_id',
			'display_mode',
			'display_active',
			'fan.speed',
			'memory.total',
			'memory.used',
			'memory.free',
			'utilization.gpu',
			'utilization.memory',
			'temperature.gpu',
			'temperature.memory',
			'power.draw',
			'power.limit',
			'clocks.gr',
			'clocks.mem'
		]
		exec(`nvidia-smi --query-gpu=${query.join(',')} --format=csv,nounits,noheader`, (error, stdout) => {
			if (error) {
				return reject(error)
			}

			const gpus: GPU[] = compact(
				stdout
					.trim()
					.split(/\n/)
					.map((line) => {
						if (!line.trim()) {
							return undefined
						}
						const raw = line
							.trim()
							.split(', ')
							.map((v) => (v.includes('N/A') ? undefined : v))
						const bus = raw[2]?.split(':')
						return {
							index: Number(raw[0]),
							name: raw[1],
							domain: bus?.[0],
							busId: bus?.[1],
							device: bus?.[2].split('.')[0],
							function: bus?.[2].split('.')[1],
							subDeviceId: isNil(raw[3]) ? undefined : Number(raw[3]),
							displayMode: raw[4] === 'Enabled',
							displayActive: raw[5] === 'Enabled',
							fanSpeed: isNil(raw[6]) ? undefined : Number(raw[6]),
							memoryTotal: isNil(raw[7]) ? undefined : Number(raw[7]),
							memoryUsed: isNil(raw[8]) ? undefined : Number(raw[8]),
							memoryFree: isNil(raw[9]) ? undefined : Number(raw[9]),
							gpuUsage: isNil(raw[10]) ? undefined : Number(raw[10]),
							memoryUsage: isNil(raw[11]) ? undefined : Number(raw[11]),
							gpuTemp: isNil(raw[12]) ? undefined : Number(raw[12]),
							memoryTemp: isNil(raw[13]) ? undefined : Number(raw[13]),
							powerDraw: isNil(raw[14]) ? undefined : Number(raw[14]),
							powerLimit: isNil(raw[15]) ? undefined : Number(raw[15]),
							gpuClock: isNil(raw[16]) ? undefined : Number(raw[16]),
							memoryClock: isNil(raw[17]) ? undefined : Number(raw[17])
						}
					})
			)

			resolve(gpus)
		})
	})
```

This calls take between 60 to 90ms to get information from your GPUs.  You can get it even faster if you reduce the amount of queries being done.  Works perfectly in our system now.  Getting it in XML format definitely slows things down in SI, so I used CSV instead.

**Comment by sebhildebrandt** [comment]
@mboudreau I guess the thing is that I am gathering a lot of information when calling `graphics()` ... Most of them are from Windows WMI and this actually needs powershell as the old `wmic` command is deprecated. For sure, if I only need information from `nvidia-smi` this command would be much faster. For the next upcoming version I am planning to split up graphics (displays and graphicCards) ... this at lease would help speed things up a little more as there is no need to retrieve `displays` every few seconds. This is a more static information. Same would be true for graphic cards but the utilization is what is interesting and therefore needs to be captured more often ... So as this is a breaking change, this will then be part of V6.

**Comment by sebhildebrandt** [comment]
@mboudreau @TheRedfoox @BlockCnFuture ... closing it for now. For V6 we will split display and GPU controller into two separate calls....
